source/d3d8to9_device.cpp

```diff
@@ -51,6 +51,10 @@ ULONG STDMETHODCALLTYPE Direct3DDevice8::AddRef()
 }
 ULONG STDMETHODCALLTYPE Direct3DDevice8::Release()
 {
+	// Shaders are destroyed alongside the device that created them in D3D8 but not in D3D9
+	// so we Release all the shaders when the device releases to mirror that behaviour
+	ReleaseShaders();
```

Comment:
I'm a bit unhappy with this, as it means shaders are destroyed even when an app does something like this, which should technically be fine:
```c++
device->CreateVertexShader();
device->AddRef();
device->Release();
device->SetVertexShader(); // Oops, shader handle is broken now
```
But since the D3D9 shader objects will hold a device reference, the `LastRefCount == 0` condition would never be triggered if an attempt was made to instead release shaders in there.
A potential solution (albeit very hacky) might be this:
```c++
ULONG STDMETHODCALLTYPE Direct3DDevice8::Release()
{
    ULONG LastRefCount = ProxyInterface->Release();

    if (LastRefCount == (VertexShaderHandles.size() + PixelShaderHandles.size())
    {
        ProxyInterface->AddRef();
        ReleaseShaders();
        LastRefCount = ProxyInterface->Release();
        assert(LastRefCount == 0);
    }

    if (LastRefCount == 0)
        delete this;

    return LastRefCount;
}
```

Reply:
If tracking the vertex declaration as per below, that condition would change to `if (LastRefCount == (VertexShaderHandles.size() * 2 + PixelShaderHandles.size())`.

Reply:
Good point, I hadn't thought of that use case. Unfortunately I think the solution is even hackier, since not all D3D9 CreateVertexDeclaration calls are matched to a D3D9 CreateVertexShader call (because no function pointer can legally be passed to use the fixed function pipeline), the reference count delta for each call to D3D8 CreateVertexShader can be 1 or 2. I added some manual reference counting for Vertex Shaders and Declarations on the D3D8to9 end to accommodate this.